astropy/io/ascii/cds.py

```diff
@@ -164,14 +207,344 @@ def get_cols(self, lines):
                     raise ValueError(f'Line "{line}" not parsable as CDS header')
 
         self.names = [x.name for x in cols]
-
         self.cols = cols
 
+    def __splitFloatFormat(self, value):
+        """
+        Splits a Float string into different parts to find number
+        of digits after decimal and check if the value is in Scientific
+        notation.
+
+        Parameters
+        ----------
+        value : the float value
+
+        Returns
+        -------
+        fmt: (int, int, int, bool, bool)
+            List of values describing the Float sting.
+            (size, dec, ent, sign, exp)
+            size, length of the given string.
+            ent, number of digits before decimal point.
+            dec, number of digits after decimal point.
+            sign, whether or not given value signed.
+            exp, is value in Scientific notation?
+        """
+        regfloat = re.compile(r"""(?P<sign> [+-]*)
+                                  (?P<ent> [^eE.]+)
+                                  (?P<deciPt> [.]*)
+                                  (?P<decimals> [0-9]*)
+                                  (?P<exp> [eE]*-*)[0-9]*""",
+                             re.VERBOSE)
+        mo = regfloat.match(value)
+
+        if mo is None:
+            raise Exception(value + " is not a float number")
+        return (len(value),
+                len(mo.group('ent')),
+                len(mo.group('decimals')),
+                mo.group('sign') != "",
+                mo.group('exp') != "")
+
+    def __set_column_val_limits(self, col):
+        """
+        Sets the `col.min` and `col.max` column attributes,
+        taking into account columns with Null values.
+        """
+        col.max = max(col)
+        col.min = min(col)
+        if col.max is np.ma.core.MaskedConstant:
+            col.max = None
+        if col.min is np.ma.core.MaskedConstant:
+            col.min = None
+
+    def columnFloatFormatter(self, col):
+        """
+        String formatter for a column containing Float values.
+        """
+        # maxSize: maximum length of string containing the float value.
+        # maxEnt: maximum number of digits places before decimal point.
+        # maxDec: maximum number of digits places after decimal point.
+        # maxPrec: maximum precision of the column values, sum of maxEnt and maxDec.
+        maxSize, maxPrec, maxEnt, maxDec = 1, 0, 1, 0
+        sign = False
+        fformat = 'F'
+        #fmt = [0, 0, 0, 0, False]
+
+        # find max sized value in the col
+        for rec in col:
+            # skip null values
+            if rec is None:
+                continue
+
+            # find format of the Float string
+            fmt = self.__splitFloatFormat(str(rec))
+
+            # if value is in Scientific notation
+            if fmt[4] is True:
```

Comment:
This is essentially a counter-measure for calculating format and width of columns containing values both in Float (Decimal format) and the Scientific notation. Like in the column `s` of the following table.

| names | e | d | s | i |
| :--------- | -- | -- | -- | -- |
| HD81809  | 1E-7 | 22.25608 | +2 | 67 |
| HD103095 | -31.6e5 | +27.2500 | -9E34 | -30 |

- First, by default, the Float column is designated as having the decimal notation, `fformat = 'F'`
- Then we go through the column, record by record, to find the exact format and the width required for it.
- If, upon splitting the Float string, it's found that the current value is in the Scientific notation, the column format designation is changed to `'E'`.
- If the previous value was in the Decimal format, the column format would have been `'F'`. If yes, then the maxSize, maxPrec and maxDec variables are set to default as well. This is done so that the column formatting happens according the value in the Scientific notation, regardless of any other values in the Decimal format.
- If the current value is not in Scientific notation and the column is marked with `'E'` (which means one of the previous values was in the Scientific notation), this particular record is simply skipped, in order to give formatting precedence to the value in Scientific notation.

All this helps correctly writing the above table as,

```
HD81809  1e-07  22.25608   2e+00  67
HD103095 -3e+06 27.25000  -9e+34 -30
```

With the ByteByByte being,

```
 Bytes Format Units  Label     Explanations
--------------------------------------------------------------------------------
 1- 8   A8     ---    names    Description of names             
10-14   E5.1   ---    e       [-3160000.0/0.01] Description of e
16-23   F8.5   ---    d       [22.25/27.25] Description of d    
25-31   E7.1   ---    s       [-9e+34/2.0] Description of s     
33-35   I3     ---    i       [-30/67] Description of i       
```

Reply:
But we do *know* the format - we write the data ourselves. We get to decide if we output in scientific notation or not - the relevant `Outputter` of the class that does the writing will format the numbers. I wonder if we should make use of this knowledge instead of turning it to strings and then inspecting the string to find what function we used to make that substitution.

On the other hand, I also wonder if this is the best use of our time. Since this seems to work, maybe we should move on to other steps (e.g. coordinates) and come back to possible simplify this in the end.

Reply:
Note that this is the expected behaviour, so that code block is not failing strictly speaking.

If any column has atleast one value in exponential notation, the intent is to convert all other values of the column to this format. The number of decimal places in the integer part is governed by the number of digits after decimal of the original value given in exponential notation. Trailing zeros after decimal are discounted.
Thus your example above. `t['b'] = [2.123456789, 2.00e20]` would also output the table.

If it is required that the output table has a set number of digits after decimal for all column values, then the original value in exponential notation should have digits, without trailing zeroes, after the decimal in the integer part. So,

```
t['b'] = [2.123456789, 2.45e20]
ascii.write(t, format='cds')

================================================================================
Byte-by-byte Description of file: table.dat
--------------------------------------------------------------------------------
 Bytes Format Units  Label     Explanations
--------------------------------------------------------------------------------
 1- 8  E8.3   ---    b       [2.12/2.45e+20] Description of b
--------------------------------------------------------------------------------
Notes:
--------------------------------------------------------------------------------
2.12e+00
2.45e+20
```

Reply:
This would not be the expected behavior for most users. Silently dropping precision based on an undocumented / tricky rule like this will eventually lead to a bug report if this gets widely used. Consider the very different behavior of two apparently similar cases:
```
In [18]: t['b'] = [2.123456789, 2.45e15]
    ...: ascii.write(t, format='cds')
    ...: 
Title:
Authors:
Table:
================================================================================
Byte-by-byte Description of file: table.dat
--------------------------------------------------------------------------------
 Bytes Format Units  Label     Explanations
--------------------------------------------------------------------------------
 1-26  F26.9  ---    b       [2.12/2450000000000000.0] Description of b
--------------------------------------------------------------------------------
Notes:
--------------------------------------------------------------------------------
               2.123456789
2450000000000000.000000000

In [19]: t['b'] = [2.123456789, 2.45e16]
    ...: ascii.write(t, format='cds')
    ...: 
Title:
Authors:
Table:
================================================================================
Byte-by-byte Description of file: table.dat
--------------------------------------------------------------------------------
 Bytes Format Units  Label     Explanations
--------------------------------------------------------------------------------
 1- 8  E8.3   ---    b       [2.12/2.45e+16] Description of b
--------------------------------------------------------------------------------
Notes:
--------------------------------------------------------------------------------
2.12e+00
2.45e+16
```

Reply:
Ah,  I see. Thank you for pointing out this discrepancy in the behaviour.
This can probably be addressed by opting for highest precision.
Would be good to address it separately. So, **needs an issue**.